Correctly handle spaces in option-values

goerz opened the pull request:
This fixes a bug in the parsing of keys with "advanced" options. For example, the following would be a valid entry in an `authorized_keys` file:

```
command="sh -c \"mysqldump db1 -u fred1 -p\"",no-x11-forwarding ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDDN2n2kt99/aYpPbOZRJeGV62R2R1h3b8CMX8fpcGGXHJAxFWg/Sz8qXcbOeTRKdnBIWlUrkDi/7hWK
```

The values for the `command` option may contain spaces, and also escaped quotes. As stated in `man sshd 8`:

> The options (if present) consist of comma-separated option
> specifications. No spaces are permitted, except within double quotes.

and

> command="command"
>     Specifies that the command is executed whenever this key is used
>     for authentication. [...]
>     A quote may be included in the command by quoting it with a
>     backslash.

In the current version of `sshkeys`, reading the above key would result in an
Exception,

```
ValueError: Could not parse key, too many fields
```

The parser has been rewritten to account for the presence of spaces and escaped quotes. The behavior is tested in a new test, `test_pubkey_line_with_quote_options`.

In addition to the above fix, the version information of the package is now defined in a `__version__` module attribute, in accordance with PEP 0396 and PEP 0386. The `setup.py` script obtains the version by parsing `sshkeys/__init__.py`.

goerz committed:
Define __version__ in module

PEP 0396 recommends that modules define a __version__ attribute and that
setup.py obtains the version by parsing that definition

goerz committed:
Test definition of __version__

The test ensures that the __version__ module attribute exists, and that
it can be parsed by distutils

goerz committed:
Merge branch 'version-number' into dev

This patch moves the definition of the version to the __version__ module
attribute

goerz committed:
Handle SSH-key options with spaces

The options given for an SSH key may contain spaces (inside a quoted
string), and escaped quotes. As stated in `man sshd 8`:

> The options (if present) consist of comma-separated option
> specifications. No spaces are permitted, except within double quotes.

and

> command="command"
>     Specifies that the command is executed whenever this key is used
>     for authentication. [...]
>     A quote may be included in the command by quoting it with a
>     backslash.

The current patch fixes a big where options containing spaces or escaped
quotes were not handled correctly, resulting in an Exception:

    ValueError: Could not parse key, too many fields

goerz committed:
Test of advanced SSH-key options

goerz committed:
Use legacy syntax for string "split" method

In Python 3, we may split a string s into at most 3 fields, separated by
whitespaces, with

    s.split(maxsplit=2)

Unfortunately, in Python 2, the split method does not take keyword
arguments, and the above has to be expressed as

    s.split(None, 2)

Since we want to be compatible with both Python 2 and Python 3, this
patch replaces the Python 3 call with the legacy version. Since this
syntax is much less readable, the logic is explained in a comment.

goerz committed:
Merge branch 'advanced-options' into dev

This patch fixes a bug when SSH keys are defined using options with values that
contain spaces.

goerz commented:
I would be happy to have this pull request accepted and released as version 0.5, if you have a chance. I'd want to use `sshkeys` in another project where keys with the command option are used extensively, so it would be great to have a new version with the bugfix installable from PyPI

mbr commented:
Guess that's what I get for being too lazy to write a parser. Thank you!

I'll give this a check, but I don't see why it couldn't be merged if it is correct. The only issue I am having with it is the version-detection in setup.py, I normally use a different way of setting these version strings (but I forgot to add it into sshkeys). Would you want me to revert your changes, cherry pick parts from the patch (which would drop you as a commit author =() or just resubmit the pull request without the version-string patches?

mbr commented on sshkeys/__init__.py:
```diff
@@ -54,50 +56,93 @@ def readable_fingerprint(self):
+        elif len(fields) == 3:
+            type_str, data64, comment = fields
+        else:
+            raise ValueError("Too many fields in key: %s"
```
I believe this error message is wrong, since maxsplit=2, this could only be raised if `len(fields)` is <= 1?

goerz replied:
See below, `maxsplit=2` produces up to 3 fields

goerz replied:
You're absolutely correct! Sorry, I completely misunderstood your comment at first. This is fixed now.

goerz committed:
Revert "Merge branch 'version-number' into dev"

This reverts commit b6ac9fd6874d59e7330c6bfdfac9fcc7d5e808b8, reversing
changes made to 7b36ca49d09c6261b4196f4edbf5475868ff22c4.

goerz committed:
Clarify maxsplit parameter in str split method

Added a comment to clarify a question to came up during code review.
When giving the `maxsplit` parameter in the `split` method of a string,
it specifies not the maximum length of the resulting list, but the
number of splits that are performed. Doing n splits results in n+1
items.

>>> s = "1 2 3 4"
>>> s.split()
['1', '2', '3', '4']
>>> s.split(maxsplit=2)
['1', '2', '3 4']

goerz commented:
I reverted the merge of the version-number change.

The maxsplit parameter is a bit confusing. Doing n splits results in n+1 items

~~Therefore, I believe the code is correct. I've added a comment in the code to clarify this.~~
EDIT: Sorry, I completely misunderstood your comment. You're completely right, of course!
This is now fixed, see below.

goerz committed:
Add test for command option with backslashes

goerz committed:
Fix bug where backlashes were dropped from options

While backslashes may be used to escape quotes in options, they should
be kept as backslashes if they're not followed by a quote. The following
is a perfectly valid line in an authorized_keys file:

    command="echo \"\HELLO\"" ssh-rsa AAAAB...

in order to print \HELLO upon connection. Currently, the backslash
before HELLO would have been dropped by the parser.

goerz commented:
I actually just found a bug in my parser, where backslashes that were not an escape for a quote would be dropped. Added two commits with a fix of this bug, and another test.

goerz committed:
Fix handling of invalid keys

Consistently raise a ValueError with an appropriate and informative
error message. The following error conditions are caught:

* empty key
* insufficient number of fields (key must contain at least a type, and
  the data, separated by a space)
* invalid data (failure of b64decode)

This also fixes a wrong error message "too many fields" in earlier
versions. There is no "too many fields" condition anymore (since
comments are allowed to have spaces). Having additional fields befor the
data field in a key will result in an "invalid data" error instead.

goerz committed:
Test exceptions for malformed keys

goerz committed:
Add .gitignore file

https://raw.githubusercontent.com/github/gitignore/master/Python.gitignore

goerz commented:
I fixed the bug with the wrong "too many fields" error message. Thanks for pointing this out, and sorry I didn't understand your comment at first. I also made sure that various malformed keys consistently produce a ValueError with an appropriate error message, with a new test.

mbr commented:
I've merged what I'm willing to merge, that is everything except the version string part (which I'll fix myself) and the `.gitignore` (Stuff like that belongs into `~/.gitgignore` in my opinion, I use the per-project ones only for really project-specific stuff).

I've also managed to preserve the authorship credit by rebasing =). Give it a look and if you're satisfied, I'll have a go at cutting a new release.

goerz commented:
Looks good!

mbr commented:
0.5 is up on PyPI.

goerz commented:
Awesome, thanks!
